### PR TITLE
Checkboxs toggle items in non-tags field lists, addresses #1957

### PR DESCRIPTION
**Pull request by inmysocks**
This extends the checkbox widget so that it can toggle items in lists in fields other than tags in the same way as in the tags field.
The code is just thrown together so there may be some cleaning to do.

### CODE REVIEW

**Comment by tobibeer** [comment]
Not sure why but the title reference doesn't link back to #1957, so I'll do it like this.

**Comment by tobibeer** [comment]
@inmysocks, I think you should amend your PR description just a little with the basic details of how you actually implemented it, e.g. via param `tagField`, I presume.

Not sure about the semantics but the convention could just be that if you use `tag` AND `field` as params then field is automatically treated as a tagField, so we don't need another parameter.

**Comment by matabele** [comment]
Works well -- may be used to toggle values in a list in any field -- like so:

```
<$list filter="[list[!!mylist]]" variable="item">
<$checkbox tag=<<item>> tagField="any-field"><<item>></$checkbox>
</$list>
```

Duplication of code appears unnecessary -- the field attribute can default to "tags" -- retaining default behaviour for backward compatibility.

**Comment by felixhayashi** [comment]
Thanks @inmysocks for this implementation and the demo!

> why, e.g. via param tagField

I understand why @inmysocks used the `tags` field as it is clearly a toggle action, closely related to the tagging operation.

But I had rather went a different path and added an attribute flag that denotes whether TW should use the `field` attribute in list mode instead of interpreting it as single-value-field. E.g. `mode="list"`. This is less confusing than a new attribute `tagField` (which is also a bit misleading) and it doesn't force the user to add the toggle value via the `tag` attribute. Instead, the toggle value could be added using the `selected` attribute. Semantically, this makes more sense to me.

-Felix

**Comment by matabele** [comment]
@inmysocks 

Using your scheme makes sense -- duplication of code can then be avoided by defaulting the 'field' attribute to 'tags' -- then using this code to manipulate the tags. This would, however, require that the 'mode' attribute defaulted to 'list' -- and to toggle a field, this default would need to be set to 'single' :-/

**Comment by matabele** [comment]
I have posted a revised version with the unnecessary duplicate code removed [here:](http://checkbox-list.tiddlyspot.com/)

Needs refinement.

**Comment by matabele** [comment]
I'm unsure of the purpose of the `field`, `checked` and `unchecked` attributes -- better functionality is available with the `$radio` widget -- as in this example:

```
<$list filter="[list[]]" variable="value">
<$radio field="modifier" value=<<value>>> <<value>></$radio>
</$list>
```

This will throw up options to select any of the alternatives resulting from the filter expression (in this case the list of titles in the `list` field.)

Dispensing with this duplicated functionality in the `$checkbox` widget would simplify extending this widget in other directions -- for example to manipulate lists in data dictionaries. This would also eliminate the conflict between the `field` and `list` attributes, when pointing to the same field.

**Comment by tobibeer** [comment]
> Dispensing with this duplicated functionality in the $checkbox widget would simplify extending this widget in other directions -- for example to manipulate lists in data dictionaries.

see #2103 ...it does not consider any list handling in data but enables the checkbox widget to work with data fields.

> Using your scheme makes sense -- duplication of code can then be avoided by defaulting the field attribute to tags -- then using this code to manipulate the tags. This would, however, require that the mode attribute defaulted to list -- and to toggle a field, this default would need to be set to single :-/

The default mode would depend on whether we specify a **tag** (list-mode) or **field** (value-mode) and then allow to override the latter to being list-mode, i.e. equivalent to tags, but addressing a different field.

I very much advocate **mode** and, starting from whatever modes we provide, to clearly define rules on how each mode works. Using a field in list-mode will surely always either set or remove a given value from the list. We should not have an attribute that says **invertTag*** but rather just one that says "invert". We should actually stop using tag specific lingo at all... tags are just one case of a list field.

**Comment by matabele** [comment]
@tobibeer

I agree -- but changing `invertTag` to `invert` will break backward compatibility :-( Would also be nice to change `tag` to `item` -- again to promote the idea that `tags` are only an instance of a list.

Perhaps the `invert` attribute value should also specify an alternate value -- any item within a list could then be toggled between two values. The `field`, `checked` and `unchecked` attributes could then be removed, as they only address the special instance of a list of 1 item (I haven't checked if parseStringArray handles this case correctly.)

My [version]http://checkbox-list.tiddlyspot.com/) already uses the same code to manipulate any list in any field, with the field defaulting to `tags` (I wish to extend this to handle data index entries as well.) The major change necessary was to generalise the hastag() function to handle any list, rather than only the `tags` field.

**Comment by matabele** [comment]
OK -- uploaded [new version.] (http://checkbox-list.tiddlyspot.com/)

Can now toggle any item in a list in any field, optionally exchanging the item for an alternate item. Works for single item lists, rendering the `field` attribute redundant (therefore, removed.) Works in all cases with `invertTag` attribute.

Thoughts?

**Comment by tobibeer** [comment]
I think what is important is to document (/and to first specify) specific modes and then clearly instruct on which attribute configurations yield which results... of course, keeping things backward compatible, i.e. tag specific attributes alongside something like **invertTag**.

I think there should be three modes:

* tag (specific for toggling tags)
* value (specific for toggling non-list fields)
* list (specific for toggling list-field-items)

...to each of which we see what attributes to associate and if we're missing any, then we create new attributes specific to a given mode so as to prevent confusion. The tag attribute can seamlessly be mapped to list mode, I think. For any fancy capabilities beyond what is possible with tags today, we can avail that through list mode. So, if a user wanted to use the same complex'ish behavior for tags, they'd be using list mode with the tags field... to avoid duplication ...and to keep things rather simple with tags. Of course, if that only complicated things for no good reason, we don't have to artificially constrain any toggle-tag capabilities.

Not sure if this will impinge on performance. Clearly, we're not using all attributes all the time. So, checking for changed attributes should only marginally slow things down, as first of all, we're just checking the existence of properties to objects.

**Comment by tobibeer** [comment]
@matabele, could you perhaps make a quick list as to how your above code / demo implements different "modes" (...regardless of whether or not we have or need a specific "mode" attribute)?

**Comment by matabele** [comment]
@tobibeer 

I have included an hasItem() function -- a replacement for hasTag() which works with any field. Therefore, the same code can be used for a list in any field (the tags field being only the default list.)

The original widget uses a `tag` attribute to specify the tag to toggle -- associated with this is the invertTag, which reverses the checkbox. I have added to this an `alt` tag, which, when specified causes the `tag` to be switched with `alt` (rather than removing it only.) This works both with an without the `invertTag` attribute.

The code handles lists of 1 item -- so no change is required to the attributes when handling a list of 1 item, or handling a list with >1 item. This renders the `field` attribute and its associated `checked`, `unchecked` and `default` attributes redundant. 

In summary:

1. To toggle a tag:  `<$checkbox tag="A Tag"> ...`
2. To inverse toggle a tag: `<$checkbox tag="A Tag" invertTag="yes"> ...`
3. To switch a tag:  `<$checkbox tag="A Tag" alt="Alternate Tag"> ...`
4. To inverse switch a tag: `<$checkbox tag="A Tag" alt="Alternate Tag" invertTag="yes"> ...`
5. To manipulate a field rather than `tags`: `<$checkbox tiddler="Tiddler to Manipulate" list="list-to-manipulate" tag="A Tag" ... and the rest is the same for manipulating tags
6. This works for lists of 1 item, without any changes.

**Comment by matabele** [comment]
@tobibeer 

Looking at the code -- perhaps the `invertTag` value, when specified, should be `true` rather than `yes`. And, of course `invertTag` should rather be `invert` (if permitted to abandon backward compatibility.)

My `list` attribute could be renamed `field` -- this would improve backward compatibility (but not maintain it entirely.) Options can be added to allow specification of the needed parameters using the old attributes, which are overridden by the newly named attributes if also specified.

**Comment by tobibeer** [comment]
> The code handles lists of 1 item -- so no change is required to the attributes when handling a list of 1 item, or when handling a list with >1 item. This renders the field attribute and its associated checked, unchecked and default attributes redundant.

This appears to be false. For a list field we have to list-encode titles with strings. For any other field we **must not** do that.

**Comment by tobibeer** [comment]
> Another option is to call the new widget something like $checklist, and include this alongside the legacy widget -- then users have time to change over their code before the $checkbox widget is depreciated at some stage.

This may actually be the cleanest approach of all.

**Comment by felixhayashi** [comment]
I think it makes sense to continue the discussion here (https://github.com/Jermolene/TiddlyWiki5/issues/1957) as this is a specific PR…

**Comment by matabele** [comment]
I believe this PR can be withdrawn. I have modified the code considerably and removed anything not related to the handling of lists.

The replacement PR may be found here: https://github.com/Jermolene/TiddlyWiki5/pull/2156

**Comment by inmysocks** [comment]
Ok, sounds good to me.

**Comment by felixhayashi** [comment]
Thanks @inmysocks for creating a prototype, it really spawned a fruitful discussion.
